pkg/server/misc.go

```diff
@@ -0,0 +1,751 @@
+package server
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/rancher/harvester/pkg/generated/clientset/versioned"
+	"github.com/rancher/harvester/pkg/server/sb"
+	"github.com/rancher/harvester/pkg/util"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	LonghornSystemNameSpace    = "longhorn-system"
+	LonghornBackendServiceName = "longhorn-backend"
+	LonghornBackendServicePort = 9500
+)
+
+var (
+	Version   = "v0.0.0-dev"
+	GitCommit = "HEAD"
+)
+
+type BundleState string
+
+const (
+	BundleStateInProgress       = BundleState("InProgress")
+	BundleStateReadyForDownload = BundleState("ReadyForDownload")
+	BundleStateError            = BundleState("Error")
+)
+
+type BundleError string
+
+const (
+	BundleErrorMkdirFailed        = BundleError("Failed to create bundle file directory")
+	BundleErrorZipFailed          = BundleError("Failed to compress the support bundle files")
+	BundleErrorOpenFailed         = BundleError("Failed to open the compressed bundle file")
+	BundleErrorStatFailed         = BundleError("Failed to compute the size of the compressed bundle file")
+	BundleProgressPercentageYaml  = 20
+	BundleProgressPercentageLogs  = 40
+	BundleProgressExternalBundle  = 60
+	BundleProgressNodeBundle      = 80
+	BundleProgressPercentageTotal = 100
+)
+
+type SupportBundleResource struct {
+	PodID              string      `json:"podID"`
+	State              BundleState `json:"state"`
+	Name               string      `json:"name"`
+	ErrorMessage       BundleError `json:"errorMessage"`
+	ProgressPercentage int         `json:"progressPercentage"`
+}
+
+type SupportBundle struct {
+	Name               string
+	State              BundleState
+	Size               int64
+	Error              BundleError
+	Filename           string
+	ProgressPercentage int
+	CreatedAt          string
+}
+
+type BundleMeta struct {
+	ProjectName          string `json:"projectName"`
+	ProjectVersion       string `json:"projectVersion"`
+	KubernetesVersion    string `json:"kubernetesVersion"`
+	ProjectNamespaceUUID string `json:"projectNamspaceUUID"`
+	BundleCreatedAt      string `json:"bundleCreatedAt"`
+	IssueURL             string `json:"issueURL"`
+	IssueDescription     string `json:"issueDescription"`
+}
+
+type SupportBundleInitateInput struct {
+	IssueURL    string `json:"issueURL"`
+	Description string `json:"description"`
+}
+
+//Support Bundle Resource
+func toSupportBundleResource(podID string, sb *SupportBundle) *SupportBundleResource {
+	return &SupportBundleResource{
+		PodID:              podID,
```

Comment:
To proxy the request to the pod which generates the bundle.
For now, I didn't implement the proxy part because the service pod is bound to the client IP.

Reply:
this is not relevant anymore.